lib/matplotlib/colors.py

```diff
@@ -1225,6 +1254,876 @@ def reversed(self, name=None):
+            x_0 = np.linspace(1, 0, lutshape[0])[:, np.newaxis] * np.ones(lutshape)
+        if not inverted[1]:
+            x_1 = np.linspace(0, 1, lutshape[1])[np.newaxis, :] * np.ones(lutshape)
+        else:
+            x_1 = np.linspace(1, 0, lutshape[1])[np.newaxis, :] * np.ones(lutshape)
+
+        # we need to use shape = 'sqare' while resampling the colormap.
+        # if the colormap has shape = 'circle' we would otherwise get *outside* in the
+        # resampled colormap
+        shape_memory = self._shape
+        self._shape = 'square'
+        if transposed:
+            new_lut = self((x_1, x_0))
+            new_cmap = BivarColormapFromImage(new_lut, name=self.name,
+                                              shape=shape_memory,
+                                              origin=(self.origin[1], self.origin[0]))
+        else:
+            new_lut = self((x_0, x_1))
+            new_cmap = BivarColormapFromImage(new_lut, name=self.name,
+                                              shape=shape_memory,
+                                              origin=self.origin)
+        self._shape = shape_memory
+
+        new_cmap._rgba_bad = self._rgba_bad
+        new_cmap._rgba_outside = self._rgba_outside
+        return new_cmap
+
+    def reversed(self, axis_0=True, axis_1=True):
+        """
+        Reverses both or one of the axis.
+        """
+        r_0 = 1
+        if axis_0:
+            r_0 = -1
+        r_1 = 1
+        if axis_1:
+            r_1 = -1
+        return self.resampled((r_0, r_1))
+
+    def transposed(self):
+        """
+        Transposes the colormap by swapping the order of the axis
+        """
+        return self.resampled((None, None), transposed=True)
+
+    def with_extremes(self, *, bad=None, outside=None, shape=None, origin=None):
+        """
+        Return a copy of the BivarColormap, for which the colors for masked (*bad*)
+        valuesand if shape = 'ignore' or 'circleignore', out-of-range *outside* values,
+        have been set accordingly.
+
+        Parameters
+        ----------
+        bad : None or Matplotlib color
+            If Matplotlib color, the *bad* value is set accordingly in the copy
+
+        outside : None or Matplotlib color
+            If Matplotlib color and shape is 'ignore' or 'circleignore', values
+            *outside* the colormap are colored accordingly in the copy
+
+        shape: str 'square' or 'circle' or 'ignore' or 'circleignore'
+
+            - If 'square' each variate is clipped to [0,1] independently
+            - If 'circle' the variates are clipped radially to the center
+              of the colormap, and a circular mask is applied when the colormap
+              is displayed
+            - If 'ignore' the variates are not clipped, but instead assigned the
+              *outside* color
+            - If  'circleignore' a circular mask is applied, but the data is not
+              clipped and instead assigned the *outside* color
```

Comment:
Yes.
In practice the lookup table will need to be a rectangle in all cases (even if the `shape` is 'circle'). This is both for practical reasons (stored as 2d array) but also because the colormap needs to contain colors that are strictly speaking outside of the circle.

Consider the following example of a circular colormap of  7×7 pixels.
![image](https://github.com/user-attachments/assets/a774060c-e18d-4209-b11c-156936732b5e)
Here, the dark pixels represents the pixles inside the circle, and the white pixels are outside.
As the original coordinate is shifted to the circle, it still ends up at a pixel that is *white*. The simple solution to this is to have a colormap that is slightly larger than the circle, for example a rectangle. 
However, when we render the colormap, we want to render only the 'dark' pixels. 

In the code, this is handled by the private variable `self._lut` ← always a rectangle.
and `@property def lut(self):` ← applies a circular mask.

Because `self._lut` is always a rectangle, and the circular mask is only applied when accessing the lut externally, it is perfectly safe to change form a rectangular to s circular colormap. 

This is also why the following two colormaps can have the same lookup table. 
![image](https://github.com/user-attachments/assets/4b33e4c6-549b-46f8-bf03-e8377ece5608)